src/server/engine_shard_set.cc

```diff
@@ -100,34 +101,79 @@ bool EngineShard::DefragTaskState::IsRequired() {
 
 // for now this does nothing
 bool EngineShard::DoDefrag() {
-  // TODO - Impl!!
-  return defrag_state_.cursor > 0;
+  // --------------------------------------------------------------------------
+  // NOTE: This task is running with exclusive access to the shard.
+  // i.e. - Since we are using shared noting access here, and all access
+  // are done using fibers, This fiber is run only when no other fiber in the
+  // context of the controlling thread will access this shard!
+  // --------------------------------------------------------------------------
+
+  constexpr size_t kCursorCountRelation = 50;
+  const float threshold = GetFlag(FLAGS_mem_utilization_threshold);
+
+  auto& slice = db_slice();
+  DCHECK(slice.IsDbValid(kDefaultDbIndex));
+  auto [prime_table, expire_table] = slice.GetTables(kDefaultDbIndex);
+  //  otherwise we will end with negative invalid segment id - crashed at CI
+  CHECK(prime_table->depth() < 40)
+      << "the depth for the prime table is invalid: " << prime_table->depth();
+  PrimeTable::Cursor cur = defrag_state_.cursor;
+  const std::size_t max_iteration = slice.DbSize(kDefaultDbIndex) / kCursorCountRelation;
```

Comment:
Do I read correctly that the task run-time is linear with the size of the database. so it will traverse for 1000000/50 iterations for 1M records ?

Reply:
Actually it not - this  limitation is about how many times you would "exit" the scan, it should be match less, for my testing for a store of 10GB it was running 2 times with this factor. maybe the name isn't great?

Reply:
your whole function runs at least `max_iteration` loops. 
DoDefrag should be very fast. It should run several buckets and exit and then continue etc.
With this code we will have unresponsive backend with it being stuck for  time linear to db size. Could be many seconds.  
We need microseconds.

Reply:
so now I understand what you mean by const

Reply:
I did a measurement for the value of 50 (const value), the time it sends running the function DoDefrag is ~0us (its about 400ns), so I think this value is good.